### PR TITLE
Add experimental support for HTTP/2

### PR DESCRIPTION
**Pull request by jumaffre**
Part of #3342 

**Now ready for review.** Thanks @wintersteiger for fixing the memory leaks!

This PR adds experimental support for client requests over HTTP/2, which will eventually allow developers to build gRPC applications. The support is very much experimental for now but we have parity with most tests run in the `e2e_logging` end-to-end test (except a few tests that test request forwarding, excluded via the `no_http2()` decorator). 

The `nghttp2` library now always gets included in the enclave application and a single RPC interface always creates HTTP/2 endpoints if the `app_protocol` configuration entry is set to `HTTP2` (default is `HTTP1`). 

Of course, there's still plenty to do to support HTTP/2 as a first-class protocol (see #3342 for follow-up work) but this PR should have just enough that we can merge this into `main`.

### CODE REVIEW

File: CMakeLists.txt
Comment (jumaffre) [inline CMakeLists.txt:108]: Note: `nghttp2` is now always part of the TCB.

Comment (wintersteiger) [reply CMakeLists.txt:108]: Hm. Should we leave it optional while it's still experimental? I think I remember that there was a problem with that, so maybe that's too much of a hassle.

File: src/http/http2.h
Comment (wintersteiger) [inline src/http/http2.h:331]: Doesn't this mean that HTTP2 client requests can never fail? It's probably easy to add by just parsing `v`.

File: src/http/http2.h
Comment (wintersteiger) [inline src/http/http2.h:386]: Yes please! This might be related to the failures I saw with `dig`.

File: src/http/http2.h
Comment (wintersteiger) [inline src/http/http2.h:435]: I think this should be added to `struct NetInterface`, which would make it operator-configurable too.

File: src/http/http2_endpoint.h
Comment (wintersteiger) [inline src/http/http2_endpoint.h:208]:
```suggestion
```

File: src/http/http2_endpoint.h
Comment (wintersteiger) [inline src/http/http2_endpoint.h:185]:
```suggestion
```

File: CMakeLists.txt
Comment (jumaffre) [reply CMakeLists.txt:108]: Because we've now cut 2.x and we're not planning to backport these changes, I think it's OK to have this on by default (still under a runtime flag!) as we'll aim to have this out for 3.x anyway? I'm happy to leave it under a compile-time flag if there are other reasons I'm missing.

File: src/http/http2_types.h
Comment (wintersteiger) [inline src/http/http2_types.h:20]: Just in case: if this is the maximum size of any request, we may want that to be a setting in the config, so that operators can adjust it if necessary.

**Review by wintersteiger — APPROVED** [review]

File: src/enclave/rpc_sessions.h
Comment (jumaffre) [reply src/enclave/rpc_sessions.h:289]: Yes, this is a temporary hack to deduce the type of endpoint (HTTP/1 or HTTP/2) without having to require the operator to specify that. I will add a comment here to explain this is only temporary.

File: src/enclave/rpc_sessions.h
Comment (jumaffre) [reply src/enclave/rpc_sessions.h:496]: Good point. I'm not sure the ACME server/JWT endpoint actually support HTTP/2? If so, this can be something we do once the two implementations are merged (i.e. pick HTTP/2 or HTTP/1 based on ALPN).

File: src/enclave/rpc_sessions.h
Comment (jumaffre) [reply src/enclave/rpc_sessions.h:519]: I spent a while on this but couldn't find a neater solution. The `ClientEndpoint` returned by the function is not an `Endpoint` as we store in `sessions` (but `HTTPClientEndpoint` and `HTTP2ClientEndpoint` are!). Again, hopefully this gets simplified once the two endpoints are merged.

File: src/http/http2.h
Comment (jumaffre) [reply src/http/http2.h:78]: Item 8. in #3342

File: src/http/http2.h
Comment (jumaffre) [reply src/http/http2.h:93]: All the `nghttp_submit_...()` calls don't actually send any data and it's up to the `nghttp2_session_send()` call to do that. This function is usually called after `nghttp2_session_mem_recv()` ([see example](https://nghttp2.org/documentation/tutorial-server.html)).

File: src/http/http2.h
Comment (jumaffre) [reply src/http/http2.h:136]: Yes that's a good point. I'll leave that to item 8. in #3342.

File: src/http/http2.h
Comment (jumaffre) [reply src/http/http2.h:198]: All of this is executed on the same thread, based on the session id: https://github.com/microsoft/CCF/blob/b749361f59e18e20ef8a2c38a190229ba2f2b92b/src/enclave/tls_endpoint.h#L83-L84

File: src/http/http2.h
Comment (jumaffre) [reply src/http/http2.h:386]: I thought there was a specific message to send to the client on stream close but I believe there's isn't. This comment is about doing some cleanup on the server side but as we currently only support a single stream per session, the session will automatically be closed at this point so there's nothing to do. In other words, this shouldn't cause any issues with `dig`.

File: src/http/http2_endpoint.h
Comment (jumaffre) [reply src/http/http2_endpoint.h:159]: Yes. This work is part of merging the two endpoints (7. in #3342).

File: src/http/http2_endpoint.h
Comment (jumaffre) [reply src/http/http2_endpoint.h:208]: Will be addressed when the two endpoints are merged.

File: src/http/http2_endpoint.h
Comment (jumaffre) [reply src/http/http2_endpoint.h:185]: Ditto.

File: tests/e2e_common_endpoints.py
Comment (jumaffre) [reply tests/e2e_common_endpoints.py:163]: This is because we don't currently enforce the maximum size of messages for HTTP/2 endpoints (as per my comment above).

File: tests/infra/clients.py
Comment (jumaffre) [reply tests/infra/clients.py]: Yes that's right. We can hopefully find another way to do this but it's low priority as `CurlClient` is mostly used for tutorials anyway.

File: tests/infra/logging_app.py
Comment (jumaffre) [reply tests/infra/logging_app.py:435]: Command forwarding is not yet supports with HTTP/2 (see 4. in #3342).

File: CMakeLists.txt
Comment (jumaffre) [reply CMakeLists.txt:108]: As discussed, we will link `nghttp2` in on `main` but won't backport HTTP/2 support to 2.x.

File: CHANGELOG.md
Comment (achamayou) [inline CHANGELOG.md:18]: I think we should have either a small mention in the doc, perhaps highlighting the fact this is experimental, or something inline here explaining how to enable it and what the limitations are.

File: src/http/http2.h
Comment (achamayou) [reply src/http/http2.h:44]: Which seems ok for small enough responses?

**Review by achamayou — APPROVED** [review]

**Pull request merged**
